.circleci/integration-tests/airflow_dag_introspection.py

```diff
@@ -0,0 +1,174 @@
+import time
+from textwrap import indent
+
+from airflow import settings
+from airflow.exceptions import AirflowNotFoundException
+from airflow.hooks.base import BaseHook
+from airflow.models import Connection
+from airflow.models.taskinstance import TaskInstance
+from airflow.utils.dot_renderer import render_dag
+from airflow.utils.log.log_reader import TaskLogReader
+from airflow.utils.session import create_session
+from sqlalchemy.orm import joinedload
+
+
+def log_checker_with_retry(max_retries: int, log_container: list):
+    """Fetch task logs with retry."""
+    retries = 0
+    print(f"type is {type(log_container)}")
+    while retries < max_retries:
+        try:
+            if log_container and log_container[0] and log_container[0][0]:
+                logs = log_container[0][0][1]
+                break
+            else:
+                print("Invalid log_container structure. Unable to retrieve logs.")
+                retries += 1
+        except IndexError:
+            print("IndexError occurred. Retrying...")
+            retries += 1
+        except Exception as e:
+            print(f"An exception occurred: {e}")
+            if retries < max_retries:
+                # Wait for some time before retrying
+                time.sleep(60)
+                print("Retrying...")
+                # Increment the number of retries
+                retries += 1
+            else:
+                print("Maximum number of retries reached.")
+    return logs
+
+
+def log_checker(ti_id: str, expected: str, notexpected: str, try_number: int = 1, **context: dict):
+    """Get Task logs of a given TI, also validate expected substring,
+    we can also provide a substring which is not expected
+    """
+    time.sleep(30)
+    dagrun = context["dag_run"]
+    task_instances = dagrun.get_task_instances()
+    this_task_instance = next(filter(lambda ti: ti.task_id == ti_id, task_instances))  # ti_call
+    context["dag"].dag_id
+    context["dag_run"].run_id
+
+    def check(ti, expect, notexpect):
+        with create_session() as session:
+            ti = (
+                session.query(TaskInstance)
+                .filter(
+                    TaskInstance.task_id == ti.task_id,
+                    TaskInstance.dag_id == ti.dag_id,
+                    TaskInstance.run_id == ti.run_id,
+                    TaskInstance.map_index == ti.map_index,
+                )
+                .join(TaskInstance.dag_run)
+                .options(joinedload("trigger"))
+                .options(joinedload("trigger.triggerer_job"))
+            ).first()
+        task_log_reader = TaskLogReader()
+
+        log_container, _ = task_log_reader.read_log_chunks(ti, try_number, {"download_logs": True})
+        logs = log_checker_with_retry(10, log_container)
+        print(f"Found logs: {logs}")
+        assert notexpect not in logs  # nosec: B101
+        assert expect in logs  # nosec: B101
```

Comment:
After a second thought, as we're actually running these lines on astro cloud, should we raise another exception when these conditions aren't met instead of using assert?